cpp/comms/src/cuML_std_comms_impl.cpp

```diff
@@ -0,0 +1,417 @@
+  CUDA_CHECK(cudaMemsetAsync(recvbuff, 0, sizeof(int), stream));
+
+  allreduce(sendbuff, recvbuff, 1, MLCommon::cumlCommunicator::INT,
+            MLCommon::cumlCommunicator::SUM, stream);
+
+  cudaStreamSynchronize(stream);
+  cudaStreamDestroy(stream);
+
+  CUDA_CHECK(cudaFree(sendbuff));
+  CUDA_CHECK(cudaFree(recvbuff));
+}
+
+void cumlStdCommunicator_impl::isend(const void *buf, int size, int dest,
+                                     int tag, request_t *request) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  request_t req_id;
+  if (_free_requests.empty())
+    req_id = _next_request_id++;
+  else {
+    auto it = _free_requests.begin();
+    req_id = *it;
+    _free_requests.erase(it);
+  }
+
+  ucp_tag_t ucp_tag = (ucp_tag_t)tag;
+
+  ucp_ep_h *ep_arr = *_ucp_eps;
+  ucp_ep_h ep_ptr = ep_arr[dest];
+
+  struct ucx_context *ucp_request = ucp_isend(ep_ptr, buf, size, tag);
+
+  _requests_in_flight.insert(std::make_pair(req_id, ucp_request));
+  *request = req_id;
+
+  ucs_status_t flush_status = flush_ep(_ucp_worker, ep_ptr);
+  printf("flush_ep completed with status %d (%s)\n", flush_status,
+         ucs_status_string(flush_status));
+#else
+  std::cout << "cuML Comms not built with UCX support" << std::endl;
+#endif
+}
+
+void cumlStdCommunicator_impl::irecv(void *buf, int size, int source, int tag,
+                                     request_t *request) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  request_t req_id;
+  if (_free_requests.empty())
+    req_id = _next_request_id++;
+  else {
+    auto it = _free_requests.begin();
+    req_id = *it;
+    _free_requests.erase(it);
+  }
+
+  ucp_ep_h *ep_arr = *_ucp_eps;
+  ucp_ep_h ep_ptr = ep_arr[source];
+
+  struct ucx_context *ucp_request =
+    ucp_irecv(_ucp_worker, ep_ptr, buf, size, tag);
+
+  _requests_in_flight.insert(std::make_pair(req_id, ucp_request));
+  *request = req_id;
+#else
+  std::cout << "cuML Comms not built with UCX support" << std::endl;
+#endif
+}
+
+void cumlStdCommunicator_impl::waitall(int count,
+                                       request_t array_of_requests[]) const {
+#ifdef WITH_UCX
+  ASSERT(_ucp_worker != nullptr,
+         "ERROR: UCX comms not initialized on communicator.");
+
+  printf("Inside waitall for rank: %d\n", getRank());
+  std::vector<struct ucx_context *> requests;
+  for (int i = 0; i < count; ++i) {
+    auto req_it = _requests_in_flight.find(array_of_requests[i]);
+    ASSERT(_requests_in_flight.end() != req_it,
+           "ERROR: waitall on invalid request: %d", array_of_requests[i]);
+    requests.push_back(req_it->second);
+    _free_requests.insert(req_it->first);
+    _requests_in_flight.erase(req_it);
+  }
+
+  int done = 0;
+  for (struct ucx_context *req : requests) {
+    if (req == nullptr) {
+      printf("Encountered null request on rank %d\n", getRank());
```

Comment:
Is a null request ever valid? If not better make this an assert. The printf should be removed anyway.